### ISSUE TITLE
Error when listing files on FTP

### ISSUE THREAD

**Issue by kevingosse** [dotnet/corefx#37175]
I'm using the following code to list the contents of a directory on a FTP server:

```c#
    var request = (FtpWebRequest)WebRequest.Create("ftp://speedtest.tele2.net/");
    request.Method = WebRequestMethods.Ftp.ListDirectory;
    var response = request.GetResponse();
```

It works on 2.x, but on 3.0.0-preview4-27615-11 it throws an exception:

```
System.Net.WebException: The remote server returned an error: 229 Entering Extended Passive Mode (|||29355|)
. ---> System.Net.Sockets.SocketException: The requested address is not valid in its context.
   at System.Net.Sockets.Socket.UpdateStatusAfterSocketErrorAndThrowException(SocketError error, String callerName)
   at System.Net.Sockets.Socket.DoBind(EndPoint endPointSnapshot, SocketAddress socketAddress)
   at System.Net.Sockets.Socket.Bind(EndPoint localEP)
   at System.Net.FtpControlStream.QueueOrCreateDataConection(PipelineEntry entry, ResponseDescription response, Boolean timeout, Stream& stream, Boolean& isSocketReady)
   at System.Net.FtpControlStream.PipelineCallback(PipelineEntry entry, ResponseDescription response, Boolean timeout, Stream& stream)
   at System.Net.CommandStream.PostReadCommandProcessing(Stream& stream)
   at System.Net.CommandStream.PostSendCommandProcessing(Stream& stream)
   at System.Net.CommandStream.ContinueCommandPipeline()
   at System.Net.FtpWebRequest.TimedSubmitRequestHelper(Boolean isAsync)
   at System.Net.FtpWebRequest.SubmitRequest(Boolean isAsync)
   --- End of inner exception stack trace ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw(Exception source)
   at System.Net.FtpWebRequest.SyncRequestCallback(Object obj)
   at System.Net.CommandStream.InvokeRequestCallback(Object obj)
   at System.Net.CommandStream.Abort(Exception e)
   at System.Net.FtpWebRequest.FinishRequestStage(RequestStage stage)
   at System.Net.FtpWebRequest.GetResponse()
   at testhttp.Program.Main(String[] args) in D:\testhttp\Program.cs:line 33
```

I originally discovered the error with an internal FTP server, but I managed to reproduce it with the first public FTP server I could find (ftp://speedtest.tele2.net/). So it does not seem to be tied to a specific server.

**Comment by kevingosse** [comment]
For what it's worth, I tried compiling a debug version of System.Net.Sockets, but now it doesn't crash if I attach the debugger. So there's a good chance that it's some kind of race condition.

**Comment by kevingosse** [comment]
I was able to reduce the repro to:

            TcpClient client = new TcpClient();
            client.Connect("speedtest.tele2.net", 21);
            var socket1 = client.Client;
            IPEndPoint localEndPoint = new IPEndPoint(((IPEndPoint)socket1.LocalEndPoint).Address, 0);
            var socket = new Socket(socket1.AddressFamily, socket1.SocketType, socket1.ProtocolType);
            socket.Bind(localEndPoint);

The difference seems to be that TcpClient is bound to InterNetwork in .net core 2.x, and InterNetworkV6 in .net core 3.0

**Comment by kevingosse** [comment]
I fixed the issue locally by using the default address family in https://github.com/dotnet/corefx/blob/master/src/System.Net.Requests/src/System/Net/FtpControlStream.cs#L1146

Basically, replacing the line with:

    Socket socket = new Socket(templateSocket.SocketType, templateSocket.ProtocolType);

However I don't know if has other impacts.

**Comment by davidsh** [comment]
>It works on 2.x, but on 3.0.0-preview4-27615-11 it throws an exception:

@kevingosse Thanks for reporting the bug.  We will look into fixing it for .NET Core 3.0.

**Comment by stephentoub** [comment]
> I would guess it is a side-effect of #30036

cc: @wfurt

**Comment by davidsh** [comment]
@kevingosse 

I tried your repro on Windows and don't get an exception.

What is the OS platform you used?  Linux?  Did you also use Docker container or normal VM?

**Comment by wfurt** [comment]
I was able to reproduce it on my Ubuntu 16.04.

**Comment by wfurt** [comment]
fixed by #37194
